operator: extract controller-runtime integration into its own cell

mhofstetter opened the pull request:
Currently the Gateway API controller (and its reconcilers) is the only controller that is making use of the controller-runtime library.

To prepare its usage for other use-cases (e.g. Ingress #28911), this commit extracts the controller-runtime integration into its own cell by providing the relevant components (Manager & Scheme) to dependent cells.

The new controller-runtime cell is responsible to start the manager itself.

Dependent cells register their reconcilers with the manager and types with the scheme respectively.

mhofstetter committed:
operator: remove unused internal gateway api model

This commit removes the unused and empty model within the gateway controller.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
operator: introduce controller-runtime cell

Currently the Gateway API controller (and its reconcilers)
is the only one that is making us of the controller-runtime library.

To prepare its use for other use-cases, this commit extracts the
controller-runtime integration into its own cell by providing the
relevant components (Manager & Scheme) to dependent cells.

The new controller-runtime cell is responsible to start the manager itself.

Dependent cells add their reconcilers to the manager and types to the scheme
respectively.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
operator: use job for controller-runtime manager lifecycle

Currently the manager of the controller-runtime is managed within
a plain Go routine.

This commit refactors this towards using a OneShot job provided by
the Hive framework.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
operator: refactor controller-runtime reconciler registration

This commit refactors the registration of the various Gateway API
related reconcilers to the controller-runtime manager by reducing
boilerplate code.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
operator: errorhandling when adding types to controller-runtime scheme

This commits adds proper errorhandling when adding types to the
controller-runtime scheme.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
operator: use k8s rest client config from clientset cell

With this commit, the k8s rest client config used by the controller-runtime's
manager is retrieved from the clientset hivecell. This way, the presence of a
k8s api server can be checked before initializing the manager.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

bimmlerd commented on operator/pkg/controller-runtime/cell.go:
```diff
@@ -45,10 +46,14 @@ func newScheme() (*runtime.Scheme, error) {
 }
 
-func newManager(lc hive.Lifecycle, logger logrus.FieldLogger, jobRegistry job.Registry, scope cell.Scope, scheme *runtime.Scheme) (ctrlRuntime.Manager, error) {
+func newManager(lc hive.Lifecycle, logger logrus.FieldLogger, jobRegistry job.Registry, scope cell.Scope, k8sClient client.Clientset, scheme *runtime.Scheme) (ctrlRuntime.Manager, error) {
```
(nit) with this many parameters, I think a `cell.In` struct reads more nicely

mhofstetter replied:
thanks, done

bimmlerd commented on operator/pkg/gateway-api/cell.go:
```diff
@@ -86,8 +87,16 @@ func registerReconcilers(mgr ctrlRuntime.Manager, enableSecretSync bool, secrets
+		gatewayv1alpha2.AddToScheme,
+	} {
+		if err := t(scheme); err != nil {
+			return fmt.Errorf("failed to add types to scheme: %w", err)
```
(nit) maybe `failed to add type %t to scheme: %w` if there's some way to get a human readable name of the type?

(also applies to the registration in `cell.go`)

mhofstetter replied:
Would be nice, but since its only a function the output of `%T` would be that much of help `func(*runtime.Scheme) error`

The underlying error will contain enough information, and these type of errors are rare anyway.

pippolo84 replied:
What about:

```go
for gv, f := range map[fmt.Stringer]func(s *runtime.Scheme) error{
	gatewayv1.GroupVersion:       gatewayv1.AddToScheme,
	gatewayv1beta1.GroupVersion:  gatewayv1beta1.AddToScheme,
	gatewayv1alpha2.GroupVersion: gatewayv1alpha2.AddToScheme,
} {
	if err := f(scheme); err != nil {
		return fmt.Errorf("failed to add types from %s to scheme: %w", gv, err)
	}
}
```

Anyway, not a blocker, I'll leave this to you.

mhofstetter replied:
Thanks for the idea @pippolo84 . This works if the package provides the group version as global variable. This is not the case for `clientgoscheme` (aggregation of multiple groupversions) that gets added to the scheme in the `controller-runtime` cell. IMO this is why we should rely on the underlying error to provide enough information for analysing any potential issue.

Update: it might work by using the `clientgoscheme.Scheme` as default scheme. This way we don't have to explicitly register this scheme.

mhofstetter replied:
Changed

bimmlerd commented on operator/pkg/gateway-api/cell.go:
```diff
@@ -60,6 +60,32 @@ func registerController(lc hive.Lifecycle, config gatewayApiConfig, ctrlRuntimeM
+
+	for _, r := range reconcilers {
+		if err := r.SetupWithManager(mgr); err != nil {
+			return fmt.Errorf("failed to setup reconciler: %w", err)
```
(nit) similarly, here, can the reconciler have a human readable name so that the error says which reconciler failed to be setup?

mhofstetter replied:
I thought about the same. The reconciler has a name, but unfortunately we're re-using the names between different reconcilers - so they aren't of that much help either.

Will follow up on this in a separate PR.

bimmlerd commented on operator/pkg/controller-runtime/cell.go:
```diff
@@ -0,0 +1,74 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package controllerruntime
```
is the mismatch between directory and pkg name intentional?

mhofstetter replied:
yes it's somehow intentional. It's more or less the same way in the library itself -> https://github.com/kubernetes-sigs/controller-runtime/blob/main/alias.go#L17

bimmlerd commented on operator/pkg/gateway-api/cell.go:
```diff
@@ -38,29 +41,27 @@ func (r gatewayApiConfig) Flags(flags *pflag.FlagSet) {
 }
 
-func registerController(lc hive.Lifecycle, config gatewayApiConfig) error {
+func registerController(lc hive.Lifecycle, config gatewayApiConfig, ctrlRuntimeManager ctrlRuntime.Manager, scheme *runtime.Scheme) error {
```
(nit) probably also time for a `cell.In`.

mhofstetter replied:
thanks, done

bimmlerd commented on CODEOWNERS:
```diff
@@ -373,6 +373,7 @@ Makefile* @cilium/build
 /netlify.toml @cilium/ci-structure
 /operator/ @cilium/operator
 /operator/pkg/ciliumenvoyconfig @cilium/sig-servicemesh
+/operator/pkg/controller-runtime @cilium/sig-servicemesh
```
I think this commit can be squashed with the commit introducing the pkg

mhofstetter replied:
agree -> squashed

bimmlerd reviewed (approved):
nits, but LGTM if you squash the CODEOWNERs commit into the commit introducing the pkg.

mhofstetter committed:
operator: use params struct (cell.In) for ctrl-runtime & gw-api cells

With this commit, params struct are used in the controller-runtime &
gateway-api cells.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

pippolo84 reviewed (approved):
LGTM, thanks :100:

meyskens commented on operator/pkg/controller-runtime/cell.go:
```diff
@@ -0,0 +1,93 @@
+// Cell integrates the components of the controller-runtime library into Hive.
+var Cell = cell.Module(
+	"controller-runtime",
+	"Manages the controller-runtime integration and its components",
```
nit: could we make a bit more useful description here? Who's controller runtime? Where to use? Here on in the comment above

mhofstetter replied:
Good idea! I changed the comment a little bit - more or less adding a link to their github project.

I don't want to add anything about _where to use it_. Currently there's no guideline that we want to use the controller-runtime for all controllers in the Cilium Operator.

meyskens reviewed (approved):
Small nits overall LGTM
